### PR TITLE
Use `tracing` spans to trace the entire MIR interp stack

### PR DESCRIPTION
**Pull request by oli-obk**
r? @RalfJung 

While being very verbose, this allows really good tracking of what's going on. While I considered schemes like the previous indenter that we had (which we could get by using the `tracing-tree` crate), this will break down horribly with things like multithreaded rustc. Instead, we can now use `RUSTC_LOG` to restrict the things being traced. You could specify a filter in a way that only shows the logging of a specific frame.

screenshot of command line output of the new formatting

If we lower the span's level to `debug`, then in `info` level logging we'd not see the frames, but in `debug` level we would see them. The filtering rules in `tracing` are super powerful, but  I'm not sure if we can specify a filter so we do see `debug` level events, but *not* the `frame` spans. The documentation at https://docs.rs/tracing-subscriber/0.2.10/tracing_subscriber/struct.EnvFilter.html makes me think that we can only turn on things, not turn off things at a more precise level.

cc @hawkw

### CODE REVIEW

**Comment by davidbarsky** [comment]
> While I considered schemes like the previous indenter that we had (which we could get by using the tracing-tree crate), this will break down horribly with things like multithreaded rustc.

I think I can get `tracing-tree` to not be a nightmare when used with concurrent programs if I introduce a fine-grained, per-span (or per-span parentage hierarchy) locking behavior.

I also think it should be possible to get something that can be dynamically switched between `tracing_subscriber::fmt::Layer` and `tracing_tree::Layer` using the [reload layer](https://docs.rs/tracing-subscriber/0.2.10/tracing_subscriber/reload/index.html). I can talk a bit more about that, if you'd like.

**Comment by oli-obk** [comment]
Oh interesting! So we could turn on `tracing-tree` if we're running without concurrency (or when another flag is set that just disables it even without concurrency)?

> locking behavior.

Uh, so... logs would not show until a thread is done with a span? So the entire span content would be buffered? That seems ok to me in most cases. We just need to turn all that off when debugging threading issues.

**Comment by Mark-Simulacrum** [comment]
I think I would personally prefer that we not try to introduce locks or anything into rustc itself but rather have a separate tool that parses and re-shuffles logger output -- is that feasible? Too hard?

**Comment by hawkw** [comment]
> The filtering rules in `tracing` are super powerful, but I'm not sure if we can specify a filter so we do see `debug` level events, but _not_ the `frame` spans. The documentation at https://docs.rs/tracing-subscriber/0.2.10/tracing_subscriber/struct.EnvFilter.html makes me think that we can only turn on things, not turn off things at a more precise level.

You should be able to use a filter like `debug,[frame]=off` (although I've not tried it). If that doesn't work, let me know — the filter implementation is quite complex and we're still trying to work out some of the kinks with more complex filtering.

**Comment by oli-obk** [comment]
>  is that feasible? Too hard?

I was considering that, too. Shouldn't be too hard to just prepend all log messages with a thread index (only if rustc is configured to use threads), then you can easily filter with just ripgrep.

**Comment by oli-obk** [comment]
> If that doesn't work, let me know — the filter implementation is quite complex and we're still trying to work out some of the kinks with more complex filtering.

I tried `RUSTC_LOG=rustc_mir::interpret=debug,[frame]=off`, but I'm seeing `frame{some::path::stuff}` in the output (with the present PR, no other local changes)

**Comment by Mark-Simulacrum** [comment]
rustc always uses threads for LLVM, at least, so it probably makes sense to just always include a thread ID.

**Comment by hawkw** [comment]
> > If that doesn't work, let me know — the filter implementation is quite complex and we're still trying to work out some of the kinks with more complex filtering.
> 
> I tried `RUSTC_LOG=rustc_mir::interpret=debug,[frame]=off`, but I'm seeing `frame{some::path::stuff}` in the output (with the present PR, no other local changes)

Ugh, that's a bummer...I guess the "static" (target-based) filter takes priority. As I mentioned, the filtering code is pretty hairy...we need to make some changes so it behaves more reasonably with complex filters.

> I was considering that, too. Shouldn't be too hard to just prepend all log messages with a thread index (only if rustc is configured to use threads), then you can easily filter with just ripgrep.

https://docs.rs/tracing-subscriber/0.2.10/tracing_subscriber/fmt/format/struct.Format.html#method.with_thread_ids :)

**Comment by davidbarsky** [comment]
@oli-obk:

> Oh interesting! So we could turn on tracing-tree if we're running without concurrency (or when another flag is set that just disables it even without concurrency)?

Maybe?

> Uh, so... logs would not show until a thread is done with a span? So the entire span content would be buffered? That seems ok to me in most cases. We just need to turn all that off when debugging threading issues.

I think so!

@Mark-Simulacrum:

> I think I would personally prefer that we not try to introduce locks or anything into rustc itself but rather have a separate tool that parses and re-shuffles logger output.

Fair enough, that's a good point!

> is that feasible? Too hard?
I suppose there are a few options: 
- Revive https://github.com/tokio-rs/console
- Build something similar to https://github.com/jonathanj/eliottree.
- Direct folks use something like [Jaeger](https://www.jaegertracing.io/) in a container with https://docs.rs/tracing-opentelemetry/0.5.0/tracing_opentelemetry/, but that might be a bit heavyweight.

In either case, I'll hold off on any changes. Do you mind commenting on https://github.com/tokio-rs/tracing/issues/884 with your needs and requirements?

**Comment by RalfJung** [comment]
So regarding this concrete PR, that's very interesting! But I feel like having *the entire stacktrace* on every line is way too verbose. Consider what this will look like in Miri where stacks are *at least* 10 frames high from the libstd `start` stuff, and can reach 50 frames easily.

But I agree that our traces could be made easier to read. Some random ideas:
* Increase indentation by 2 for each frame, but in a way that actually helps. I know there is existing code for that but at least for Miri output I never visually "saw" this, it just wasn't helpful. I am not sure why. I think part of it is that Miri's logging is entirely separate from the core engine logging, and doesn't indent with it -- is that fixable with `tracing`?
* Print the full stack trace occasionally. Like maybe each time it changes (function call/return)?

File: src/librustc_mir/interpret/eval_context.rs [hunk 1]
```diff
@@ -48,8 +48,38 @@ pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
...
+        if let Some(id) = span.id() {
+            log::dispatcher::get_default(|d| {
+                d.enter(&id);
+            });
+        }
+        // This executes the destructor on the previous instance of `SpanGuard`, ensuring that
+        // we never enter or exit more spans than vice versa. Unless you `mem::leak`, then we
+        // can't protect the tracing stack, but that'll just lead to weird logging, no actual
+        // problems.
+        *self = Self(span);
```
Comment (RalfJung) [inline src/librustc_mir/interpret/eval_context.rs:68]: Wow this is super confusing and IMO needs more comments.^^

So the thing we do here is that `SpanGuard` calls "enter" on something in `log::dispacher` inside `SpanGuard::enter`, and it calls `exit` on drop. But we call the exit *after* the enter, and somehow it still works? And what if we create a `SpanGaurd` and never `enter` it, then we `exit` something that was never entered?

It doesn't help that `Span` is already something very different in rustc so my brain just makes all the wrong connections all the time.^^

File: src/librustc_mir/interpret/eval_context.rs [hunk 2]
```diff
@@ -48,8 +48,38 @@ pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
         FxHashMap<(Ty<'tcx>, Option<ty::PolyExistentialTraitRef<'tcx>>), Pointer<M::PointerTag>>,
 }
 
+struct SpanGuard(log::span::Span);
```
Comment (hawkw) [inline src/librustc_mir/interpret/eval_context.rs:51]: I don't _entirely_ get the motivation behind this type versus just using the `tracing` `Span` type directly...why is this needed? Are there potential upstream changes we can make that would make this type unnecessary?

File: src/librustc_mir/interpret/eval_context.rs [hunk 3]
```diff
@@ -48,8 +48,38 @@ pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
...
+impl SpanGuard {
+    fn new() -> Self {
+        Self(log::span::Span::none())
+    }
+
+    fn enter(&mut self, span: log::span::Span) {
+        if let Some(id) = span.id() {
+            log::dispatcher::get_default(|d| {
+                d.enter(&id);
+            });
```
Comment (hawkw) [inline src/librustc_mir/interpret/eval_context.rs:62]: This is a slightly different behavior than what `tracing`'s `Span` type does, and (potentially) subtly incorrect if multiple subscribers are in use. 

The `Span` struct [tracks the dispatcher that observed the span's creation](https://github.com/tokio-rs/tracing/blob/0da8e3f37636845ece6bdceeecb8a0946cc77076/tracing/src/span.rs#L363-L367) (the call to `new_span`) and gave it an `Id`. Subsequently, when the span is entered, exited, or dropped, it [ensures that those methods are called on the original subscriber](https://github.com/tokio-rs/tracing/blob/0da8e3f37636845ece6bdceeecb8a0946cc77076/tracing/src/span.rs#L716-L718), even if it has moved to the context of a _different_ subscriber. 

This is because span ID assignment is only guaranteed to be unique on a per-subscriber basis. A different subscriber may have assigned the same ID to an unrelated span. If multiple subscribers are in use, this could result in a messed-up trace. 

I think we could implement the same behavior as `tracing::Span` here if we used [`Span::with_subscriber`](https://docs.rs/tracing/0.1.18/tracing/span/struct.Span.html#method.with_subscriber), like this:
```suggestion
        span.with_subscriber(|(id, dispatch)| {
            dispatch.enter(id);
        });
```

File: src/librustc_mir/interpret/eval_context.rs [hunk 4]
```diff
@@ -48,8 +48,38 @@ pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
...
+    }
+}
+
+impl Drop for SpanGuard {
+    fn drop(&mut self) {
+        if let Some(id) = self.0.id() {
+            log::dispatcher::get_default(|d| {
+                d.exit(&id);
+            });
+        }
```
Comment (hawkw) [inline src/librustc_mir/interpret/eval_context.rs:78]: Similarly to my above comment, I think this needs to be:
```suggestion
        self.0.with_subscriber(|(id, dispatch)| {
            dispatch.exit(id);
        });
```

File: src/librustc_mir/interpret/eval_context.rs [hunk 5]
```diff
@@ -48,8 +48,38 @@ pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
...
+    }
+}
+
 /// A stack frame.
-#[derive(Clone)]
```
Comment (hawkw) [inline src/librustc_mir/interpret/eval_context.rs:52]: `tracing::Span` is `Clone`, so I think this type _could_ continue to derive clone...

Comment (RalfJung) [reply src/librustc_mir/interpret/eval_context.rs:52]: I agree with this diff though, there's no reason we should clone a stack frame.

**Comment by oli-obk** [comment]
I played around a bit with `tracing-tree`, what do you think of something like ? That doesn't repeat the current stack every now and then, but it makes it fairly easy to track. I could also add something at the closing of a span. This way we basically get back today's behaviour with the start/end logs of a stack frame, but we also have the tree structure

File: src/librustc_mir/interpret/eval_context.rs (hunk 5 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs:52]: yea this is more of a drive-by fix

File: src/librustc_mir/interpret/eval_context.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs:51]: I want do be able to automatically exit the span on `Drop`. When I use `.enter()` I get a guard, but this guard borrows the `Span`, so I can't store both of them in the same datastructure. I could probably manually invoke `enter` and `exit` at the push/pop of the stack frames, but that seems fragile compared to using a guard like this.

File: src/librustc_mir/interpret/eval_context.rs (hunk 3 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs:62]: ooh, that makes sense.

**Comment by RalfJung** [comment]
Hm... that looks promising yes. Is there some way to cap this so that when we have 70 nested frames we don't have a terminal full of vertical bars? Maybe after depth 50, reset to 0 and start over with the nesting or so?

**Comment by oli-obk** [comment]
> Is there some way to cap this so that when we have 70 nested frames we don't have a terminal full of vertical bars? Maybe after depth 50, reset to 0 and start over with the nesting or so?

I'll have a look, that should be doable

**Comment by oli-obk** [comment]
Note that with the appropriate filters we can hide arbitrary levels. So we could set it up that the logs only start after a specific frame has been pushed and stop after that frame is popped. We'd could build a list of useful filters for ppl to pass to `RUSTC_LOG`.

File: src/librustc_mir/interpret/eval_context.rs (hunk 2 above)
Comment (hawkw) [reply src/librustc_mir/interpret/eval_context.rs:51]: > I want do be able to automatically exit the span on `Drop`. When I use `.enter()` I get a guard, but this guard borrows the `Span`, so I can't store both of them in the same datastructure.

Ah, okay, that makes sense. A change we could make upstream to better support this use case is adding a method to `Span` that's like `enter`, but takes the span by value, and returns an owned guard. The owned guard would have an explicit `exit` method that returns the unentered span, as well as exiting on drop. I discussed this proposal here: https://github.com/tokio-rs/tracing/issues/79#issuecomment-503276300

This would be quite simple to implement upstream. Thus far, we haven't done so just because there doesn't seem to be a whole lot of demand for it, but this seems like a good use-case.

Comment (hawkw) [reply src/librustc_mir/interpret/eval_context.rs:51]: Another note: this type should probably not implement `Send`. Entering a span applies to the _current_ thread; when a span is entered, the thread on which `enter` is called is considered to be in that span. While the `Span` type can be freely sent or shared across threads, the entry should not — if this type is sent to a different thread and then dropped, it will call `exit` on _that_ thread, which may have never entered the span; meanwhile, the thread where the span _was_ entered will remain in that span forever.

**Comment by oli-obk** [comment]
All the features discussed in this PR have been merged into `tracing_tree`, here's a picture of how the logs look with colors and other formatting:

Screenshot from 2020-08-11 12-21-42

In textual form:

```
2:rustc INFO rustc_mir::interpret::eval_context PAUSING
2:rustcrustc_mir::interpret::eval_context::frame{message=unwind::libunwind::_Unwind_Exception::private::{{constant}}#0}
2:rustc├─0ms  INFO rustc_mir::interpret::eval_context ENTERING
2:rustc├─0ms  INFO rustc_mir::interpret::step _0 = const unwind::libunwind::unwinder_private_data_size
2:rustc├─0ms  INFO rustc_mir::interpret::eval_context PAUSING
2:rustc├─rustc_mir::interpret::eval_context::frame{message=unwind::libunwind::unwinder_private_data_size}
2:rustc│ ├─0ms  INFO rustc_mir::interpret::eval_context ENTERING
2:rustc│ ├─0ms  INFO rustc_mir::interpret::step _0 = const 6_usize
2:rustc│ ├─0ms  INFO rustc_mir::interpret::step return
2:rustc│ ├─0ms  INFO rustc_mir::interpret::eval_context LEAVING(0) unwind::libunwind::unwinder_private_data_size (unwinding = false)
2:rustc├─0ms  INFO rustc_mir::interpret::step return
2:rustc├─0ms  INFO rustc_mir::interpret::eval_context LEAVING(0) unwind::libunwind::_Unwind_Exception::private::{{constant}}#0 (unwinding = false)
```

File: src/librustc_mir/interpret/eval_context.rs (hunk 1 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs:71]: Yea, the ordering wasn't ideal (I changed it), but also technically irrelevant as `enter` never gets called twice, but if it were, this would be sane.

I documented all this on the `new´ and `enter` methods

**Comment by RalfJung** [comment]
I like that rendering, except for this part
```
rustc_mir::interpret::eval_context::frame{message=unwind::libunwind::unwinder_private_data_size}
```
Is there any way to make the function name more prominent here? I feel it drowns a bit in the noise, with the `{message=...}` around it.

Also the `ENTERING` is rather redundant, isn't it? (LEAVING is nice because it repeats the function name.)

**Comment by RalfJung** [comment]
Also PAUSING doesnt have the function name any more (not sure why) and looks very different than the other messages? It has no timing and the "INFO" is before the module instzad of after.

And what about `CONTINUING`?

File: src/librustc_mir/interpret/eval_context.rs [hunk 6]
```diff
@@ -48,8 +48,41 @@ pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
         FxHashMap<(Ty<'tcx>, Option<ty::PolyExistentialTraitRef<'tcx>>), Pointer<M::PointerTag>>,
 }
 
+// The Phantomdata exists to prevent this type from being `Send`. If it were sent across a thread
```
Comment (RalfJung) [inline src/librustc_mir/interpret/eval_context.rs:51]:
```suggestion
// The `PhantomData` exists to prevent this type from being `Send`. If it were sent across a thread
```

File: src/librustc_mir/interpret/eval_context.rs [hunk 7]
```diff
@@ -651,7 +688,11 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         self.frame_mut().locals = locals;
 
         M::after_stack_push(self)?;
-        info!("ENTERING({}) {}", self.frame_idx(), self.frame().instance);
+
+        let span = info_span!("frame", "{}", instance);
+        self.frame_mut().tracing_span.enter(span);
```
Comment (RalfJung) [inline src/librustc_mir/interpret/eval_context.rs:693]: Is there some way we can avoid rightwards drift with stacks of 50 functions and more?

Looks like the indentation is done fully automatically by tracing so this might be hard to control.

**Comment by RalfJung** [comment]
Also my question is still open of how this looks like in Miri, when there are also Miri log messages between the core engine log messages.

File: src/librustc_mir/interpret/eval_context.rs (hunk 7 above)
Comment (davidbarsky) [reply src/librustc_mir/interpret/eval_context.rs:693]: There is: https://github.com/davidbarsky/tracing-tree/pull/15. This is in the most recent release of tracing-tree on crates.io.

**Comment by oli-obk** [comment]
> with the `{message=...}` around it.

oops, I didn't notice that. I'm gonna remove all that. This wrapping isn't necessary for hierarchical printing

File: src/librustc_mir/interpret/eval_context.rs (hunk 7 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/eval_context.rs:693]: Oh that's awesome. :)

Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs:693]: this is already avoided :) after a depth of 10 we wrap around. Example: 
https://user-images.githubusercontent.com/2067774/89716739-8c86b800-d97d-11ea-9b74-78f040b845aa.png

File: src/librustc_driver/lib.rs [hunk 8]
```diff
@@ -1239,11 +1239,19 @@ pub fn init_env_logger(env: &str) {
...
-    builder.init()
+    let filter = tracing_subscriber::EnvFilter::from_env(env);
+    let layer = tracing_tree::HierarchicalLayer::default()
+        .with_indent_lines(true)
+        .with_ansi(true)
+        .with_targets(true)
+        .with_thread_ids(true)
+        .with_thread_names(true)
+        .with_wraparound(10)
+        .with_indent_amount(2);
```
Comment (RalfJung) [inline src/librustc_driver/lib.rs:1250]: Oh wow these are some global decisions being made here.^^ Maybe that should be a separate PR, it's not really related to the interpreter stack.

I'd probably set wrap-around to 25 or so. Most screens today are fairly wide.
Everything else is the default I assume? Might be worth saying that in a comment.

File: src/librustc_mir/interpret/eval_context.rs (hunk 7 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/eval_context.rs:693]: Hm, could there be some indication like a "..." to show that wrap-around happened?

**Comment by oli-obk** [comment]
> Also PAUSING doesnt have the function name any more (not sure why) and looks very different than the other messages?

imo `PAUSING` is now redundant, too, you can just scroll up and see what block you are in.

File: src/librustc_driver/lib.rs (hunk 8 above)
Comment (oli-obk) [reply src/librustc_driver/lib.rs:1250]: > Oh wow these are some global decisions being made here.^^ Maybe that should be a separate PR, it's not really related to the interpreter stack.

We need something here, otherwise there won't be any wrapping or indent. I don't really want to drag out this PR even longer. If ppl disagree with the default settings they can open issues.

> I'd probably set wrap-around to 25 or so. Most screens today are fairly wide.

it's counted in indentations, not in chars, so that's would be 50 chars, which is half your terminal. Anyway, we can fine tune this later or make it configurable via env vars, or...

> Everything else is the default I assume?

No, everything being configured is not the default.

**Comment by RalfJung** [comment]
> > Also PAUSING doesnt have the function name any more (not sure why) and looks very different than the other messages?
> 
> imo `PAUSING` is now redundant, too, you can just scroll up and see what block you are in.

That could be hundreds of lines further up, for large functions.

File: src/librustc_driver/lib.rs (hunk 8 above)
Comment (RalfJung) [reply src/librustc_driver/lib.rs:1250]: Uh, but then why so many changes from the default?^^

> it's counted in indentations, not in chars, so that's would be 50 chars, which is half your terminal. Anyway, we can fine tune this later or make it configurable via env vars, or...

Fair.

**Comment by oli-obk** [comment]
>That could be hundreds or even thousands of lines further up, for large functions.

true... maybe we could add a flag to tracing-tree that automatically does this (repeat current span, mention new span, ... stuff ..., mention current span, pop span, mention now active span)

**Comment by RalfJung** [comment]
> > with the `{message=...}` around it.
> 
> oops, I didn't notice that. I'm gonna remove all that. This wrapping isn't necessary for hierarchical printing

@oli-obk what is the status of this?

I also still think we should repeat the function name in all of these markers, including PAUSING.

**Comment by oli-obk** [comment]
> I also still think we should repeat the function name in all of these markers, including PAUSING.

this is solved :stuck_out_tongue: The opening and closing of an event will now display the span information (see the example in https://github.com/davidbarsky/tracing-tree/blob/main/examples/basic.stdout), so we don't need to manually do that via events

This is now just waiting on https://github.com/davidbarsky/tracing-tree/pull/19 getting merged and @davidbarsky publishing a new version after that

**Comment by RalfJung** [comment]
> this is solved stuck_out_tongue The opening and closing of an event will now display the span information (see the example in https://github.com/davidbarsky/tracing-tree/blob/main/examples/basic.stdout), so we don't need to manually do that via events

Ah, neat!

I find the output a bit confusing though... I don't know how to read it.
```
1:main│└┐basic::server{host="localhost", port=8080}
1:main│ ├─ms INFO basic starting
1:main│ ├─ms INFO basic listening
1:main│ ├┐basic::server{host="localhost", port=8080}
1:main│ │└┐basic::conn{peer_addr="82.9.9.9", port=42381}
1:main│ │ ├─ms DEBUG basic connected
```
The second `basic::server` here is intended one more than the first, so I am not sure if they refer to the same thing. I would have expected something more like
```
1:main│└┐basic::server{host="localhost", port=8080}
1:main│ ├─ms INFO basic starting
1:main│ ├─ms INFO basic listening
1:main│ │basic::server{host="localhost", port=8080}
1:main│ ├┐basic::conn{peer_addr="82.9.9.9", port=42381}
1:main│ │├ms DEBUG basic connected
```

**Comment by oli-obk** [comment]
> Ideally with an extra marker for whenever `basic::server` is repeated, indicating that this is _not_ a new thing but a repetition of old data.

The extra marker is `┐` ;)

If you come up with better symbols, I'm very open to other things, but I'd rather get something merged and bikeshed symbols in the tracing-tree bugtracker

**Comment by RalfJung** [comment]
> If you come up with better symbols, I'm very open to other things, but I'd rather get something merged and bikeshed symbols in the tracing-tree bugtracker

Sure.

**Comment by davidbarsky** [comment]
> This is now just waiting on davidbarsky/tracing-tree#19 getting merged and @davidbarsky publishing a new version after that.

Sorry about that! Totally slipped my mind—I read the notification on GitHub and left it in my tabs. Taking care of that now; I'll update y'all when I publish a new version in about 30 minutes.

**Comment by oli-obk** [comment]
No worries! Thanks for taking care of it now.

**Comment by davidbarsky** [comment]
Done, @oli-obk! https://github.com/davidbarsky/tracing-tree/releases/tag/v0.1.6

**Comment by oli-obk** [comment]
New output (modulo the `PAUSING`, that is gone in the latest push)

Screenshot from 2020-09-28 19-55-27

**Comment by RalfJung** [comment]
Yes that rendering looks great. :)

File: compiler/rustc_mir/src/interpret/eval_context.rs [hunk 9]
```diff
@@ -824,12 +855,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         }
 
         if !self.stack().is_empty() {
-            info!(
-                "CONTINUING({}) {} (unwinding = {})",
-                self.frame_idx(),
-                self.frame().instance,
-                unwinding
-            );
+            info!(unwinding);
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/eval_context.rs:858]: Why is `info!(unwinding)` run twice? Also just having `unwinding=true/false` in the log without much context is not very clear...

Comment (oli-obk) [reply compiler/rustc_mir/src/interpret/eval_context.rs]: The duplication is gone, and we now inform about the stack popping again

Ooh, maybe we should try out `#[instrument]` next :laughing:

Comment (RalfJung) [reply compiler/rustc_mir/src/interpret/eval_context.rs]: > Ooh, maybe we should try out #[instrument] next laughing

No idea what that means.^^
